### PR TITLE
separate RTL and Mission mode

### PR DESCRIPTION
**Pull request by KonradRudin**
### Solved Problem
Clean up to separate the RTL mode from the mission mode. Currently when in RTL, part of the RTL logic is executed within the mission mode.

### Solution

- Add a base class which loads the mission items and starts and stops the mission mode
- Refactor mission mode to separate it from the reverse and fast forward execution mode. It only executes the planned mission.
- Add separate classes for RTL mission reverse and mission fast forward. Those classes can interact directly with the planned mission instead of relying on the mission mode.

Bug Fixes in the process:

- RTL for mission reverse is now working andand lands on the home position.
- Updated RTL logic such that for RTL type 1, if a safepoint exists, it is always preferred over the home position.

Behavior changes:
- Mission now updates the dataman cache only when it is active, this could lead to a blocked calculation state on the activation of the mode. Once activated, it should not block anymore as before.
 - Calculation fro the land item is now performed on mission upload instead afterwards in the mission module.
 - Dataman Client has additional function for writing. While a write is performed the mission item is also updated in the cache if it is loaded.

### Test coverage
- Unit/integration test: Performed both SIH and SITL testing for VTOL and MC. Mission including RTL and VTL landing patterns, also include jump items.
- Added MAVSDK test for RTL mission and MIssion reverse.
- Memory testing: increase memory usage by ~1.5K. Mostly due to multiple mission base classes loaded at teh same time.
- Flight tests with VTOL and a fmu_v6x: 
https://review.px4.io/plot_app?log=ac832f39-3a08-4e43-b320-748ac6936c2b
https://review.px4.io/plot_app?log=e75a9931-eaf5-4687-a0bc-598c6b55e910
https://review.px4.io/plot_app?log=25f890f8-8a6f-46e3-a14c-0a8c59370c5b

### Context
Related links, screenshot before/after, video

### CODE REVIEW

**Review by RomanBapst — DISMISSED** [review]

File: test/mavsdk_tests/test_vtol_mission.cpp [hunk 1]
```diff
@@ -44,3 +44,46 @@ TEST_CASE("Fly VTOL mission", "[vtol]")
...
+	AutopilotTesterRtl tester;
+	tester.connect(connection_url);
+	tester.wait_until_ready();
+	tester.set_takeoff_land_requirements(0);
+	tester.load_qgc_mission_raw_and_move_here("test/mavsdk_tests/vtol_mission_without_landing.plan");
+	// vehicle should follow the mission in reverse and land at the home position
+	tester.set_rtl_type(2);
+	tester.arm();
+	tester.execute_rtl_when_reaching_mission_sequence(6);
+	tester.check_tracks_mission_raw(30.0f);
```
Comment (RomanBapst) [inline test/mavsdk_tests/test_vtol_mission.cpp:87]:
```suggestion
	tester.check_tracks_mission_raw(300.0f); // this check is currently disabled since there seems to be a bug in mavsdk which causes the reported mission sequence to be out of order
```

**Comment by hamishwillee** [comment]
@KonradRudin Return mode has been one of the hardest things historically do document - you can see now complicated it is here: https://docs.px4.io/main/en/flight_modes/return.html#return-types-rtl-type

When this is done, if there is any change to the external behaviour, can you help me update please?

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -2240,15 +2240,18 @@ FixedwingPositionControl::Run()
 
 				_position_setpoint_previous_valid = PX4_ISFINITE(_pos_sp_triplet.previous.lat)
 								    && PX4_ISFINITE(_pos_sp_triplet.previous.lon)
-								    && PX4_ISFINITE(_pos_sp_triplet.previous.alt);
+								    && PX4_ISFINITE(_pos_sp_triplet.previous.alt)
+								    && _pos_sp_triplet.previous.valid;
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2244]: Why is that change needed? Doesn't it break what was patched with https://github.com/PX4/PX4-Autopilot/pull/19942?

**Review by sfuhrer — COMMENTED** [review]
Out of curiosity: how much complexity (and with that flash) would we safe if we'd remove the reverse mission functionality all together?

**Comment by KonradRudin** [comment]
> Out of curiosity: how much complexity (and with that flash) would we safe if we'd remove the reverse mission functionality all together?

Hey @sfuhrer i have checked, removing the mission reverse only saves 1K Flash. Mission reverse is quite simple in the implementation.

**Review by RomanBapst — APPROVED** [review]

**Comment by RomanBapst** [comment]
@KonradRudin Let me quickly merge another PR of mine which will use about 20k of flash. Then we can merge

File: msg/Mission.msg [hunk 3]
```diff
@@ -4,6 +4,9 @@ uint8 dataman_id	# default 0, there are two offboard storage places in the datam
 uint16 count		# count of the missions stored in the dataman
 int32 current_seq	# default -1, start at the one changed latest
 
-int16 mission_update_counter # indicates updates to the mission, reload from dataman if increased
-int16 geofence_update_counter # indicates updates to the geofence, reload from dataman if increased
-int16 safe_points_update_counter # indicates updates to the safe points, reload from dataman if increased
+int32 land_start_index  # Index of the land start marker, if unavailable index of the land item, -1 otherwise
+int32 land_index 	# Index of the land item, -1 otherwise
```
[the remark this answers is not in the document]
Comment (KonradRudin) [reply msg/Mission.msg:8]: I think it should be clear that both counts as land item, right?

**Comment by KonradRudin** [comment]
@RomanBapst i squashed a few commits and rebased on newest main.

**Review by RomanBapst — DISMISSED** [review]

**Comment by tstastny** [comment]
@dagar would you mind taking a final pass? @RomanBapst and @sfuhrer already went through. But it is a large PR. So just to have another set of eyes. otherwise if you're happy - ill merge so we can get to the follow up pr with more VTOL functionality.

**Comment by dagar** [comment]
Has this been flight tested on real hardware?

File: src/modules/navigator/rtl_direct.cpp [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+using namespace math;
+
+RtlDirect::RtlDirect(Navigator *navigator) :
+	MissionBlock(navigator),
+	ModuleParams(navigator)
+{
+	_param_mpc_z_v_auto_up = param_find("MPC_Z_V_AUTO_UP");
+	_param_mpc_z_v_auto_dn = param_find("MPC_Z_V_AUTO_DN");
+	_param_mpc_land_speed = param_find("MPC_LAND_SPEED");
+	_param_fw_climb_rate = param_find("FW_T_CLMB_R_SP");
```
Comment (dagar) [inline src/modules/navigator/rtl_direct.cpp:63]: This isn't great because it marks both FW & MC parameters active in the parameter tree. A user with a regular quadcopter will go into QGC parameters and see random FW parameters.

**Comment by dagar** [comment]
Does anything prevent the mission feasibility checker from complaining twice?

**Comment by KonradRudin** [comment]
> Has this been flight tested on real hardware?

Hey @dagar  I have tested it in SITL quite a bit and in SIH on the hardware as well, and checked on SIH for the dataman client interface issues and memory issues. Not real flights though it should only affect the high level waypoint guidance law, which i think can easily be tested on SIH.

> Does anything prevent the mission feasibility checker from complaining twice?

Yes, i have added a new field in the mission_result message, which now also writes down the instance count of the mission, which was feasibility checked. See here: https://github.com/PX4/PX4-Autopilot/blob/5e2dcbd865d4685fe0db89726001cd30a55f3bf3/src/modules/navigator/mission_base.cpp#L582-L591. So a mission feasibility is only checked, if it was not yet checked for this instance count. Not thread safe, but since it currently runs sequentially, it should work, since only the first, one updates it.

**Comment by KonradRudin** [comment]
@dagar Finally i could add some flight logs to this PR you can find them in the PR description. Also here is a flight log, where we have an error report from dataman https://review.px4.io/plot_app?log=6ba0b9b1-bbdb-41f7-ae07-a21e44f6459c  don't know yet why this happened.

**Comment by KonradRudin** [comment]
@dagar Found the instances where a mission item load request could be made without initialized mission index, which might have caused the issue in the last flight test. Is fixed in the newest commit.

**Comment by KonradRudin** [comment]
Rebased onto current main again

**Comment by sfuhrer** [comment]
Rebased and solved some conflicts that came in with https://github.com/PX4/PX4-Autopilot/pull/20908. Also made one fix for the mission resume: https://github.com/PX4/PX4-Autopilot/pull/20903/commits/5895f0ed5e693b53cd23f6034ec37d949684c408.
Tested it then in SITL on quad and VTOL a bit.

**Comment by sfuhrer** [comment]
Rebased and solved some conflicts that came in with https://github.com/PX4/PX4-Autopilot/pull/20908. Also made one fix for the mission resume: https://github.com/PX4/PX4-Autopilot/commit/5895f0ed5e693b53cd23f6034ec37d949684c408.
Tested it then in SITL on quad and VTOL a bit.

**Comment by KonradRudin** [comment]
Rebased again on current main

**Comment by KonradRudin** [comment]
rebased on current main

**Comment by KonradRudin** [comment]
squashed commits

**Review by sfuhrer — APPROVED** [review]

**Comment by KonradRudin** [comment]
@hamishwillee the docs should still be up to date, so no changes necessary.
